File: exercises/day17-Santa-Gift-Dispatcher/java/src/main/java/D.java
// Decompiled / patched-up version used for the exercise.
// I *think* this was some kind of main component, but the name never got fixed.
import java.util.*;

public class D {

    // looks like some map of counts? keys are probably strings but I'm not 100% sure.
    private Map s;
    // list of "something" that gets processed later, maybe requests?
    private List r = new ArrayList();

    // pretty sure this was called early on with some initial data map (stock? config?).
    public D(Map m) {
        this.s = new HashMap();
        if (m != null) {
            for (Object k : m.keySet()) {
                this.s.put(k, m.get(k)); // just copy everything over for now
            }
        }
    }

    // adds one entry: a name plus a list of values (preferences? items?). naming is terrible.
    public void a(String x, List y) {
        if (x == null || y == null) return;
        List cp = new ArrayList();
        cp.addAll(y); // I remember wanting a defensive copy here...
        r.add(new E(x, cp));
    }

    // main method that runs through r and uses s somehow. z = max per entry? not documented.
    public List b(int z) {
        List out = new ArrayList();
        if (z <= 0) return out;

        for (int i = 0; i < r.size(); i++) {
            Object o = r.get(i);
            if (!(o instanceof E)) continue; // should not really happen but who knows.
            E e = (E) o;
            int c = z;
            while (c > 0) {
                Object picked = f(e);
                if (picked == null) break; // nothing left to grab for this one
                out.add(new G(e.n, picked.toString()));
                c--;
            }
        }
        return out;
    }

    // tries to fetch ONE thing for the given E from the map s.
    private Object f(E e) {
        if (e == null || e.w == null) return null;

        // first pass: walk through whatever is in e.w in order and see if s has it
        for (int i = 0; i < e.w.size(); i++) {
            Object wish = e.w.get(i);
            if (wish == null) continue;
            Integer cnt = (Integer) s.get(wish);
            if (cnt != null && cnt.intValue() > 0) {
                s.put(wish, Integer.valueOf(cnt.intValue() - 1)); // consume one
                return wish;
            }
        }

        // fallback: couldn't match anything in e.w, so just grab the first entry with count > 0
        Object altKey = null;
        Integer altCnt = null;
        for (Object k : s.keySet()) {
            Object v = s.get(k);
            if (v instanceof Integer) {
                Integer ci = (Integer) v;
                if (ci.intValue() > 0) {
                    altKey = k;
                    altCnt = ci;
                    break;
                }
            }
        }
        if (altKey == null || altCnt == null) return null;
        s.put(altKey, Integer.valueOf(altCnt.intValue() - 1));
        return altKey;
    }

    // tiny holder class: name + some list (looks like an ordered list of "wanted" things).
    private static class E {
        String n;
        List w;

        E(String n, List w) {
            this.n = n;
            this.w = w;
        }
    }

    // seems to represent one "output row": two strings, no idea why it's called G.
    public static class G {
        public final String a;
        public final String b;

        public G(String a, String b) {
            this.a = a;
            this.b = b;
        }
    }
}

File: exercises/day17-Santa-Gift-Dispatcher/kotlin/src/main/kotlin/D.kt
// Decompiled / patched-up version used for the exercise.
// This used to have nicer names, but they were stripped; didn't get around to fixing them.
import java.util.*

// main-ish class, some sort of dispatcher/allocator from what I remember.
class D(m: Map<*, *>?) {

    // map of something to an Int count (stock? quota?). left generic on purpose.
    private val s: MutableMap<Any?, Int> = HashMap()
    // bunch of entries that get iterated over later.
    private val r: MutableList<E> = ArrayList()

    init {
        if (m != null) {
            for (k in m.keys) {
                val v = m[k]
                if (v is Int) {
                    s[k] = v // only keep integer-like values
                }
            }
        }
    }

    // registers one "row": some identifier x with a list y (probably ordered preferences).
    fun a(x: String?, y: List<Any?>?) {
        if (x == null || y == null) return
        val cp: MutableList<Any?> = ArrayList()
        cp.addAll(y) // shallow copy, better than aliasing the original
        r.add(E(x, cp))
    }

    // does the main processing loop; z is some upper bound per E (max items per entry?).
    fun b(z: Int): List<G> {
        val out: MutableList<G> = ArrayList()
        if (z <= 0) return out

        for (e in r) {
            var c = z
            while (c > 0) {
                val picked = f(e) ?: break
                out.add(G(e.n, picked.toString()))
                c--
            }
        }
        return out
    }

    // internal helper that tries to take ONE item from s that matches this E.
    private fun f(e: E?): Any? {
        if (e == null || e.w == null) return null

        // stage one: go through e.w, in order, and see if s still has any of those
        for (wish in e.w) {
            if (wish == null) continue
            val cnt = s[wish]
            if (cnt != null && cnt > 0) {
                s[wish] = cnt - 1 // burn one
                return wish
            }
        }

        // stage two: nothing from the list worked, so take the first leftover thing in s
        var altKey: Any? = null
        var altCnt: Int? = null
        for ((k, v) in s) {
            if (v > 0) {
                altKey = k
                altCnt = v
                break
            }
        }
        if (altKey == null || altCnt == null) return null
        s[altKey] = altCnt - 1
        return altKey
    }

    // tiny data holder: some name plus a list w (seems like an ordered wishlist).
    private class E(val n: String, val w: List<Any?>?)

    // result pair: label + chosen value; original name "G" is not very helpful.
    class G(val a: String, val b: String)
}
